### PR TITLE
sysctl.d support

### PR DESCRIPTION
**Pull request by ned21**
Support for /etc/sysctl.d with unit tests.

### CODE REVIEW

File: ncm-sysctl/src/main/pan/components/sysctl/schema.pan [hunk 1]
```diff
@@ -16,8 +16,8 @@ include { 'quattor/schema' };
 
 type component_sysctl_structure = {
   include structure_component
-  
-  'command'   : string = '/sbin/sysctl'
+
+  'command'   : string = '/sbin/sysctl' with match(SELF, '^/.+')
```
Comment (stdweird) [inline ncm-sysctl/src/main/pan/components/sysctl/schema.pan:20]: this allows arbitrary commands to be executed. can we move the executable to the component? what other values are expected here?

File: ncm-sysctl/src/main/perl/sysctl.pm [hunk 2]
```diff
@@ -39,52 +39,27 @@ sub Configure {
...
-				   goodre => '\s*'.$key.'\s*=\s*'.$value,
-				   good => "$key = $value",
-				   add => 'last'
-				  );
-	if ($st < 0) {
-	    $self->error("Failed to update sysctl $key (value=$value)");
-	} else {
-	    $changes += $st;
-	}
+    if ($configFile =~ m{/}) {
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:55]: this should be 

``` perl
use Readonly;
Readonly my $SYSCTL_CONF => '/etc/sysctl.conf';
if ($configFile == $SYSCTL_CONF) {
```

ortherwise you can overwrite arbitrary files (esp in combination with executing arbitrary commands)

File: ncm-sysctl/src/main/perl/sysctl.pm [hunk 3]
```diff
@@ -39,52 +39,27 @@ sub Configure {
...
+        $self->debug(1, "confFile setting is a path");
+        $changes = $self->sysctl_file($configFile, $sysctl_exe, $variables);
+    } else {
+        $self->debug(1, "confFile is a single file, using sysctl.d");
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:59]: it's always a single file, maybe `confFile is a relative filename`?

File: ncm-sysctl/src/main/perl/sysctl.pm [hunk 4]
```diff
@@ -94,13 +69,60 @@ sub Configure {
...
                             "$output");
+        } else {
+            $self->debug(1, "$sysctl_exe output: $output");
+        }
+    }
+    return 1;
+}
+
+# new method for managing files in /etc/sysctl.d
+sub sysctl_dir {
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:84]: `{` on new line

File: ncm-sysctl/src/main/perl/sysctl.pm [hunk 5]
```diff
@@ -94,13 +69,60 @@ sub Configure {
...
+    print $fh "# Written by ncm-sysctl, do not modify\n";
+    foreach my $key (sort(keys(%$variables))) {
+        my $value = $variables->{$key};
+        print $fh "$key = $value\n";
+    }
+    return $fh->close();
+}
+
+# legacy code for managing /etc/sysctl.conf
+sub sysctl_file {
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:101]: `{` on new line

File: ncm-sysctl/src/main/perl/sysctl.pm [hunk 6]
```diff
@@ -94,13 +69,60 @@ sub Configure {
...
+    unless (-e $configFile && -w $configFile) {
+        $self->warn("Sysctl configuration file does not exist ",
+                    "or is not writable ($configFile)");
+        return;
+    }
+
+    my $changes = 0;
+    foreach my $key (sort(keys(%$variables))) {
+        my $value = $variables->{$key};
+        my $st = NCM::Check::lines($configFile,
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:112]: use `CAF::FileEditor`?

File: ncm-sysctl/src/main/perl/sysctl.pod [hunk 7]
```diff
@@ -30,15 +30,22 @@ Default : false
...
 Default : /etc/sysctl.conf
 
 
 =head2 /software/components/ncm-sysctl/variables : nlist (optional)
 
-A nlist of key/value defining sysctl variables. There is no check that the key matches a valid key, so be cautious
-to use appropriate variable names. Key name should be escaped if it begins by a non-alphabetic character.
+A nlist of key/value defining sysctl variables. There is no check that
+the key matches a valid key, so be cautious to use appropriate
+variable names. Key names must begin with a letter or an underscore.
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pod:48]: mention that values are not quoted, so if the value has whitespace, quotes must be added in the profile.

File: ncm-sysctl/src/test/perl/configure.t [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+
+my $cmp = NCM::Component::sysctl->new('sysctl');
+my $cfg = get_config_for_profile('simple');
+
+is($cmp->Configure($cfg), 1, "Configure succeeds");
+my $fh = get_file("/etc/sysctl.d/50-quattor");
+ok($fh, "A file was actually created");
+isa_ok($fh, "CAF::FileWriter");
+
```
Comment (stdweird) [inline ncm-sysctl/src/test/perl/configure.t:32]: add a `like("$fh", qr{^kernel.sysrq = 1$}m, "Found variable")`

File: ncm-sysctl/src/main/pan/components/sysctl/schema.pan (hunk 1 above)
Comment (ned21) [inline ncm-sysctl/src/main/pan/components/sysctl/schema.pan:20]: If that's something we care about. Is the intention to restrict arbitrary command execution to only limited or legacy module like filecopy?

Wouldn't a breaking change like that also need to be carried out over multiple releases? e.g. first change schema to make it optional and have the code issue a warning if its set, then remove the functionality in a future release?

File: ncm-sysctl/src/main/perl/sysctl.pod [hunk 9]
```diff
@@ -30,15 +30,22 @@ Default : false
...
 =head2 /software/components/ncm-sysctl/confFile : string (required)
 
-String defining sysctl configuration file.
+String defining sysctl configuration file. If this value contains a /
+character then it will be treated as an absolute path to a file which
+is modified in place, and a backup made.
+
+If the value does not contain a / then it will be treated as the name
+of a file to be created in /etc/sysctl.d. The existing contents of
+the file will be overwritten.
```
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pod:39]: if the file does not end with `*.conf`, i don't think it's read on reboot (`sysctl` has a `--system` option).
if possible, it might be good to enforce the `.conf` extension in the schema?

File: ncm-sysctl/src/main/perl/sysctl.pm (hunk 6 above)
Comment (ned21) [inline ncm-sysctl/src/main/perl/sysctl.pm:112]: I was going for the approach of, "If it ain't broke, and its a deprecated code path, don't risk an outage by messing with it for no reason."

File: ncm-sysctl/src/main/pan/components/sysctl/schema.pan (hunk 1 above)
Comment (stdweird) [inline ncm-sysctl/src/main/pan/components/sysctl/schema.pan:20]: do you want arbitrary executable here? maybe we can start by restricting the schema and opening an issue for deprecation in future release. on EL7, it's `/usr/sbin/sysctl`, so people should be aware of this (and the schema can maybe be `with match(SELF, '^(/usr)?/sbin/sysctl$')`)

File: ncm-sysctl/src/main/perl/sysctl.pm (hunk 6 above)
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:112]: ok, can we then add a warning/info/verbose that this deprecated? also in the pod?
feel free to open an issue for `CAF::FileEditor` to unittest `NCM::Check` vs `CAF::FileEditor` behaviour. so we can be more confident in the future when cleaning up code.

File: ncm-sysctl/src/main/perl/sysctl.pm (hunk 2 above)
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pm:55]: (`$configFile eq $SYSCTL_CONF` ofcourse)

File: ncm-sysctl/src/main/perl/sysctl.pm (hunk 6 above)
Comment (ned21) [inline ncm-sysctl/src/main/perl/sysctl.pm:112]: Yes, but I would advise against it for the first release. The new code path depends on having at least versions initscripts-8.45.38-2.el5 and initscripts-9.03.27-1.el6 installed so we cannot just default the new code. My thinking was that we make a release with the new code as optional, but the old behaviour as default. Sites where they can test in PAN that they are using a compatible RH/SL release can then opt-in to the new code. If everything works, and/or we can make a declaration that only SL releases which have the necessary initscripts version are supported, then the default could be changed in a future release, and a warning/info message included to warn those still using the old code path.

For my own part, we will be using the new code on all RH5 & 6 hosts but are stuck with the old code on RH4 so I'd prefer not to get warnings about something I cannot do anything about...

File: ncm-sysctl/src/main/perl/sysctl.pod (hunk 7 above)
Comment (ned21) [inline ncm-sysctl/src/main/perl/sysctl.pod:48]: That would actually be pretty easy to automate the addition of quotes in code where the value contains a space, but would break people who already have quoted values so I'll update the doc. :(

File: ncm-sysctl/src/main/perl/sysctl.pod (hunk 9 above)
Comment (ned21) [inline ncm-sysctl/src/main/perl/sysctl.pod:39]: In `/etc/init.d/functions`

``` bash
apply_sysctl() {
    sysctl -e -p /etc/sysctl.conf >/dev/null 2>&1
    for file in /etc/sysctl.d/* ; do
        is_ignored_file "$file" && continue
        test -f "$file" && sysctl -e -p "$file" >/dev/null 2>&1
    done
}

is_ignored_file() {
    case "$1" in
        *~ | *.bak | *.orig | *.rpmnew | *.rpmorig | *.rpmsave)
            return 0
            ;;
    esac
    return 1
}
```

So based on this, I think we could also keep a .bak copy of the old file?

File: ncm-sysctl/src/main/pan/components/sysctl/schema.pan (hunk 1 above)
Comment (ned21) [inline ncm-sysctl/src/main/pan/components/sysctl/schema.pan:20]: I want to not break anyone who's relying on this functionality. :) I think it's a good idea to put backwards incompatible changes in their own PR so they can be discussed separately to the new functionality being created here as we may need to consult a wide audience.

**Comment by ned21** [comment]
Moved to https://github.com/quattor/configuration-modules-core/pull/405

File: ncm-sysctl/src/main/perl/sysctl.pod (hunk 9 above)
Comment (stdweird) [inline ncm-sysctl/src/main/perl/sysctl.pod:39]: @ned21 i see you already closed this PR, but http://www.freedesktop.org/software/systemd/man/sysctl.d.html means `.conf` only. since there is no real benefit in the file extension, i propose to enforce `.conf` and yes, a `.bak` seems doable for backup. (it would be nice if `CAF::File*` could store backups elsewhere (new issue?))

File: ncm-sysctl/src/main/pan/components/sysctl/schema.pan (hunk 1 above)
Comment (stdweird) [inline ncm-sysctl/src/main/pan/components/sysctl/schema.pan:20]: what we really need is some tool/script where we can check new schemas with existing profiles (to see the impact of such changes).
it could be as simple as

``` pan
object templates test;

bind /some/path = new_type;
```

and compile for server `machine` with 

```
ant -Dpan.root.element="value('//$machine/')"
```
